### PR TITLE
Add ExcludedPrefixes option to VersioningConfig

### PR DESCRIPTION
**Pull request by VerdantForge**
Add ExcludedPrefixes option to VerionningConfig.

Mimicks the behaviour of the minio cli `mc` [link](https://min.io/docs/minio/linux/administration/object-management/object-versioning.html#id9)

Started from issue #1401

### CODE REVIEW

File: minio/versioningconfig.py [hunk 1]
```diff
@@ -34,16 +34,20 @@ class VersioningConfig:
     """Versioning configuration."""
 
     def __init__(
-            self,
-            status: str | None = None,
-            mfa_delete: str | None = None,
+        self,
+        status: str | None = None,
+        mfa_delete: str | None = None,
+        excluded_prefixes: list[str] | None = None,
```
Comment (balamurugana) [inline minio/versioningconfig.py:40]:
```suggestion
            self,
            status: str | None = None,
            mfa_delete: str | None = None,
            excluded_prefixes: list[str] | None = None,
            exclude_folders: bool = False,
```

File: minio/versioningconfig.py [hunk 2]
```diff
@@ -34,16 +34,20 @@ class VersioningConfig:
...
     ):
         if status is not None and status not in [ENABLED, SUSPENDED]:
             raise ValueError(f"status must be {ENABLED} or {SUSPENDED}")
         if mfa_delete is not None and mfa_delete not in [ENABLED, DISABLED]:
             raise ValueError(f"MFA delete must be {ENABLED} or {DISABLED}")
+        if excluded_prefixes is not None and not isinstance(excluded_prefixes, list):
+            raise ValueError(f"Excluded prefixes must be a list")
```
Comment (balamurugana) [inline minio/versioningconfig.py]:
```suggestion
```

File: minio/versioningconfig.py [hunk 3]
```diff
@@ -34,16 +34,20 @@ class VersioningConfig:
...
         self._status = status
         self._mfa_delete = mfa_delete
+        self._excluded_prefixes = excluded_prefixes
```
Comment (balamurugana) [inline minio/versioningconfig.py:50]:
```suggestion
        self._excluded_prefixes = excluded_prefixes
        self._exclude_folders = exclude_folders
```

File: minio/versioningconfig.py [hunk 4]
```diff
@@ -55,12 +59,25 @@ def mfa_delete(self) -> str | None:
         """Get MFA delete."""
         return self._mfa_delete
 
+    @property
+    def excluded_prefixes(self) -> str | None:
+        """Get MFA delete."""
+        return self._excluded_prefixes
+
```
Comment (balamurugana) [inline minio/versioningconfig.py:66]:
```suggestion
    @property
    def excluded_prefixes(self) -> list[str] | None:
        """Get excluded prefixes."""
        return self._excluded_prefixes

    @property
    def exclude_folders(self) -> bool:
        """Get exclude folders."""
        return self._exclude_folders

```

File: minio/versioningconfig.py [hunk 5]
```diff
@@ -55,12 +59,25 @@ def mfa_delete(self) -> str | None:
...
-        return cls(status, mfa_delete)
+
+        excluded_prefixes_tag = element.find("ExcludedPrefixes")
+        excluded_prefixes = None
+        if excluded_prefixes_tag:
+            excluded_prefixes = [
+                tag.text for tag in excluded_prefixes_tag.findall("ExcludedPrefix")
+            ]
+
+        return cls(status, mfa_delete, excluded_prefixes)
```
Comment (balamurugana) [inline minio/versioningconfig.py:80]:
```suggestion
        excluded_prefixes = [
            prefix.text for prefix in findall(
                element, "ExcludedPrefixes/Prefix",
            )
        ]
        exclude_folders = findtext(element, "ExcludeFolders") == "true"
        return cls(
            status,
            mfa_delete,
            cast(Union[List[str], None], excluded_prefixes),
            exclude_folders,
        )
```

File: minio/versioningconfig.py [hunk 6]
```diff
@@ -69,4 +86,8 @@ def toxml(self, element: ET.Element | None) -> ET.Element:
             SubElement(element, "Status", self._status)
         if self._mfa_delete:
             SubElement(element, "MFADelete", self._mfa_delete)
+        if self._excluded_prefixes:
+            SubElement(element, "ExcludedPrefixes")
+            for prefix in self._excluded_prefixes:
+                SubElement(element, "ExcludedPrefix", prefix)
```
Comment (balamurugana) [inline minio/versioningconfig.py:92]:
```suggestion
        for prefix in self._excluded_prefixes or []:
            SubElement(
                SubElement(element, "ExcludedPrefixes"), "Prefix", prefix,
            )
        if self._exclude_folders:
            SubElement(element, "ExcludeFolders", "true")
```

File: tests/unit/versioningconfig_test.py [hunk 7]
```diff
@@ -18,7 +18,11 @@
 
 from minio import xml
 from minio.commonconfig import DISABLED, ENABLED
-from minio.versioningconfig import OFF, SUSPENDED, VersioningConfig
+from minio.versioningconfig import (
+    OFF,
+    SUSPENDED,
+    VersioningConfig,
+)
```
Comment (balamurugana) [inline tests/unit/versioningconfig_test.py:25]:
```suggestion
```

File: tests/unit/versioningconfig_test.py [hunk 8]
```diff
@@ -53,3 +57,22 @@ def test_config(self):
...
+                <ExcludedPrefixes>
+                    <ExcludedPrefix>prefix</ExcludedPrefix>
+                </ExcludedPrefixes>
+            </VersioningConfiguration>
+            """,
+        )
+        xml.marshal(config)
+        self.assertEqual(config.status, ENABLED)
+        self.assertEqual(config.mfa_delete, DISABLED)
+        self.assertListEqual(config.excluded_prefixes, ["prefix"])
```
Comment (balamurugana) [inline tests/unit/versioningconfig_test.py:78]:
```suggestion
```

File: tests/unit/versioningconfig_test.py [hunk 9]
```diff
@@ -18,7 +18,6 @@
 
 from minio import xml
 from minio.commonconfig import DISABLED, ENABLED
-from minio.versioningconfig import OFF, SUSPENDED, VersioningConfig
```
Comment (balamurugana) [inline tests/unit/versioningconfig_test.py:21]: Revert this change

File: minio/versioningconfig.py
Comment (balamurugana) [inline minio/versioningconfig.py:1]: fix two imports
```
from typing import Type, TypeVar, List, Union, cast
```
and 
```
from .xml import Element, SubElement, findall, findtext
```

**Review by balamurugana — CHANGES_REQUESTED** [review]

File: tests/unit/versioningconfig_test.py (hunk 8 above)
Comment (VerdantForge) [reply tests/unit/versioningconfig_test.py]: Why remove the test?

Comment (balamurugana) [reply tests/unit/versioningconfig_test.py]: We don't need any minio server specific tests here.

Comment (balamurugana) [reply tests/unit/versioningconfig_test.py]: Not needed as we may change the specific when required.

File: minio/versioningconfig.py [hunk 10]
```diff
@@ -37,13 +37,19 @@ def __init__(
...
             mfa_delete: str | None = None,
+            excluded_prefixes: list[str] | None = None,
+            exclude_folders: bool = False,
     ):
         if status is not None and status not in [ENABLED, SUSPENDED]:
             raise ValueError(f"status must be {ENABLED} or {SUSPENDED}")
         if mfa_delete is not None and mfa_delete not in [ENABLED, DISABLED]:
             raise ValueError(f"MFA delete must be {ENABLED} or {DISABLED}")
+        if excluded_prefixes is not None and not isinstance(excluded_prefixes, list):
+            raise ValueError("Excluded prefixes must be a list")
```
Comment (balamurugana) [inline minio/versioningconfig.py:48]:
```suggestion
```

**Comment by balamurugana** [comment]
depends on https://github.com/minio/minio-py/pull/1403

**Review by balamurugana — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]
